### ISSUE TITLE
Handle image updates gracefully

### ISSUE THREAD

**Issue by JoeyZwicker** [pachyderm/pachyderm#251]
What should Pachyderm's behavior be around user images specified in a pipeline?

Pachyderm cannot control when the user may update those images on DockerHub or a private registry, which effectively is like updating the pipeline because everything could have changed.

Users should need be explicit with their image specification, declaring an explicit image/tag. We could also support floating tags such as "latest", but that's a little risky because someone can push a new "latest" and the pipeline can start churning out crappy results without anyone noticing.

**Comment by JonathanFraser** [comment]
It's funny, I was thinking about this on the way to work this morning. The solution, I think, is that code is data and data is code. I see little difference between merging data from two seperate repos and merging code with data. If you were to version your code in pachyderm, then you would get the behaviour you want implicitly I think.  I mean really, what you guys have built here is a distributed build system for data.

**Comment by jdoliner** [comment]
Versioning the code is definitely going to be the key here.
I think a pretty easy way to get the behavior we want would be the following:

When a pipeline is first created we resolve the image name to a hash. We do this only once on the node the receives the `CreatePipeline` request so there's no chance of a race. This means that `latest` tags work fine, because they'll still get resolved to a hash when the request is received.

Updating the pipeline will reresolve the image hash if an image is present in the update request. If an image changes there is no incrementality.

**Comment by jdoliner** [comment]
> I mean really, what you guys have built here is a distributed build system for data

That's actually a really good way of describing Pachyderm... I'm stealing that :stuck_out_tongue:

**Comment by sjezewski** [comment]
In our discussion yesterday, we agreed that this issue should just cover the simple fix (JD describes). If we want more advanced updating mechanisms, we'll make a new issue.

**Comment by jdoliner** [comment]
At this point I think we, and the broader k8s community, have pretty well established practices in place for updating images. I don't think the things mentioned in this issue make sense to do anymore, so I'm going to close this issue. Feel free to reopen if you feel there's something we can do here.
